### PR TITLE
kubevirtci, presubmit: add jobs to build multi-arch fedora-realtime and fedora-with-test-tooling image

### PR DESCRIPTION
**Pull request by zhlhahaha**
This is a draft of the implementation on how we use vm-image-builder image to build multi-arch fedora-realtime and fedora-with-test-tooling images.

I do not integrate the publish process into the job `publish-kubevirtci`. I only trigger the build job when related files changes.
As fedora-realtime and fedora-with-test-tooling do not change often, and it resource consuming to do the muti-arch build, it is not necessary to publish them in every release.

### CODE REVIEW

**Comment by zhlhahaha** [comment]
/hold
until the following PR get merged. 
https://github.com/kubevirt/project-infra/pull/2630
https://github.com/kubevirt/project-infra/pull/2631
https://github.com/kubevirt/kubevirtci/pull/732
And we have new vm-image-builder image.

**Comment by andreabolognani** [comment]
Most of the nuance of defining prow jobs goes way over my head, so take this with a grain of salt O:-) but I can't spot anything obviously wrong here. The logic limiting rebuilds to when the definitions have changed looks reasonable too.

Just out of curiosity, is there a mechanism through which select prow jobs can be manually kicked off by one of the admins? Or would that require pushing dummy changes to the repo?

**Comment by zhlhahaha** [comment]
> Just out of curiosity, is there a mechanism through which select prow jobs can be manually kicked off by one of the admins? Or would that require pushing dummy changes to the repo?

From my understanding, the only way to kicked off a prow job is modify the configuration and push it to the repo. @dhiller @brianmcarey How's your idea?

**Comment by zhlhahaha** [comment]
Hi, @brianmcarey no need to rehearsal now, as the image are not built yet
```
        containers:
        - image: quay.io/kubevirtci/vm-image-builder
```

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml
Comment (brianmcarey) [inline github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml:194]: This will probably break automation that we have for bumping kubevirtci in kubevirt/kubevirt - there won't be corresponding kubevirtci providers to match the tag in  the `latest` file. 

The automation for this PR - https://github.com/kubevirt/kubevirt/pull/9616

https://github.com/kubevirt/kubevirt/blob/main/hack/bump-kubevirtci.sh

**Comment by brianmcarey** [comment]
> Hi, @brianmcarey no need to rehearsal now, as the image are not built yet
> 
> ```
>         containers:
>         - image: quay.io/kubevirtci/vm-image-builder
> ```

Ah sorry - missed that. I thought it was ready for a trial run

**Comment by brianmcarey** [comment]
> > Just out of curiosity, is there a mechanism through which select prow jobs can be manually kicked off by one of the admins? Or would that require pushing dummy changes to the repo?
> 
> From my understanding, the only way to kicked off a prow job is modify the configuration and push it to the repo. @dhiller @brianmcarey How's your idea?

The prow admins have methods of kicking off jobs manually if required. We can generate the yaml for the prowjob and just apply it against our control plane cluster.

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml
Comment (zhlhahaha) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml:194]: It will not influence the current process, as it upload to file `fedora-realtime-latest` and `fedora-with-test-tooling` rather than `latest` via the following commend.
```
 gsutil cp ./latest gs://kubevirt-prow/release/kubevirt/kubevirtci/fedora-realtime-latest
```
And the script `bump-kubevirtci.sh` is checking the file `latest` in the google cloud.
https://github.com/kubevirt/kubevirt/blob/29526b6cd255c4e3c0bf13995628559655733fe4/hack/bump-kubevirtci.sh#LL8C54-L8C55

Comment (brianmcarey) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml:194]: Yes you are right - should be fine.

Comment (zhlhahaha) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml:194]: In case misunderstanding, I will re-write the lines like the following.
```
            echo $KUBEVIRTCI_TAG > fedora-realtime-latest &&
            gsutil cp ./fedora-realtime-latest gs://kubevirt-prow/release/kubevirt/kubevirtci/fedora-realtime-latest
```

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml
Comment (brianmcarey) [inline github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml:84]: This preset is not required - see `preset-podman-in-container-enabled: "true"` below

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml
Comment (brianmcarey) [inline github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml:89]: we have moved to a new workloads cluster recently called `kubevirt-prow-workloads`

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-presubmits.yaml
Comment (brianmcarey) [inline github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-presubmits.yaml:319]: cluster name -> `kubevirt-prow-workloads`

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-presubmits.yaml
Comment (brianmcarey) [inline github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-presubmits.yaml:374]: cluster name -> kubevirt-prow-workloads

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-presubmits.yaml
Comment (brianmcarey) [inline github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-presubmits.yaml:393]: anti-affinity not required here

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-presubmits.yaml
Comment (brianmcarey) [inline github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-presubmits.yaml:338]: anti-affinity not required here

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml
Comment (brianmcarey) [inline github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml:99]: anti-affinity not required here

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml
Comment (brianmcarey) [inline github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml:155]: dind preset not required here

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml
Comment (brianmcarey) [inline github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml:170]: anti-affinity not required here

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml
Comment (brianmcarey) [inline github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml:160]: cluster name -> `kubevirt-prow-workloads`

**Review by brianmcarey — COMMENTED** [review]
Looks good @zhlhahaha  - thanks !

Just a couple of comments as the environment has changed recently.

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml
Comment (zhlhahaha) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-postsubmits.yaml:84]: Ok, so preset-dind-enabled means enable Docker in Docker,
```
- labels:
    preset-dind-enabled: "true"
  env:
  - name: DOCKER_IN_DOCKER_ENABLED
    value: "true"
  volumes:
  - name: docker-graph
    emptyDir: {}
  volumeMounts:
  - name: docker-graph
    mountPath: /docker-graph
```

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-presubmits.yaml
Comment (dhiller) [inline github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-presubmits.yaml:318]: Please set

```
skip_report: true
```

for this presubmit and the presubmit below. Otherwise failed jobs will lead to `tide` not merging.

Docs: https://docs.prow.k8s.io/docs/jobs/#how-to-configure-new-jobs

(we can and should enable reporting when it's stable enough)

**Review by dhiller — CHANGES_REQUESTED** [review]
In general this all looks great, one comment inline.

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-presubmits.yaml
Comment (zhlhahaha) [reply github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-presubmits.yaml:318]: Thanks Daniel, updated.

**Comment by zhlhahaha** [comment]
Hold until https://github.com/kubevirt/kubevirtci/pull/732 get merged and a multi-arch vm-image-builder is built via https://github.com/kubevirt/project-infra/pull/2631

**Comment by brianmcarey** [comment]
@zhlhahaha looks like the rehearsals timed out 

https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_project-infra/2760/rehearsal-check-provision-fedora-with-test-tooling/1674243794909794304

**Comment by zhlhahaha** [comment]
> @zhlhahaha looks like the rehearsals timed out
> 
> https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_project-infra/2760/rehearsal-check-provision-fedora-with-test-tooling/1674243794909794304

Yes, it hangs on creating domain. I am not sure why this happen. Maybe the issue is related to the nested virtualization. It would be good if we can see the libvirt log. Can you porting the libvirtd.log? Maybe run the following command, change the podname to the rehearse job pod name.
```
kubectl exec -n kubevirt-prow-jobs podname -- cat /var/log/libvirtd.log
```

By the way, here is the build log in my local environment. It works fine but it is on baremetal machine.
```
+ virt-install --memory 2048 --vcpus 2 --name provision-vm --disk example_build/copy-example-image.qcow2,device=disk --disk cloudinit.iso,device=cdrom --os-type Linux --os-variant fedora32 --graphics none --network default --import --virt-type kvm --noautoconsole --wait 120
WARNING  --os-type is deprecated and does nothing. Please stop using it.

Starting install...
Creating domain...                                                                                       |    0 B  00:00:00     

Domain is still running. Installation may be in progress.
Waiting 120 minutes for the installation to complete.
```

**Comment by zhlhahaha** [comment]
rehearsal after https://github.com/kubevirt/kubevirtci/pull/1038 get merge.

**Review by brianmcarey — COMMENTED** [review]
@zhlhahaha can we split this PR in two? One for the presubmits and one for the post submits? 

We can merge the presubmits then which will allow you to debug the issues in a kubevirtci PR. 

What do you think ?

File: github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-presubmits.yaml
Comment (brianmcarey) [inline github/ci/prow-deploy/files/jobs/kubevirt/kubevirtci/kubevirtci-presubmits.yaml:319]: I think it is worth setting `skip_report: false` for both presubmits as it is nice to get the feedback on the PR rather than having to find the job in the prow dashboard.

**Review by brianmcarey — COMMENTED** [review]
/lgtm

/unhold

These presubmits should help with debugging the build on the workloads cluster.
Thanks @zhlhahaha

**Comment by zhlhahaha** [comment]
> These presubmits should help with debugging the build on the workloads cluster. Thanks @zhlhahaha

Yes, of course, this would be very helpful. Thank Brian.

**Pull request merged**
